backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/v2/invoicecandidates/impl/CreateInvoiceCandidatesService.java

```diff
@@ -0,0 +1,575 @@
+			@NonNull final ManualCandidateService manualCandidateService,
+			@NonNull final ExternallyReferencedCandidateRepository externallyReferencedCandidateRepository,
+			@NonNull final ProductRestService productRestService,
+			@NonNull final BPRelationsService bpRelationsService,
+			@NonNull final JsonServiceFactory jsonServiceFactory)
+	{
+		this.currencyService = currencyService;
+		this.docTypeService = docTypeService;
+		this.manualCandidateService = manualCandidateService;
+		this.externallyReferencedCandidateRepository = externallyReferencedCandidateRepository;
+		this.productRestService = productRestService;
+		this.bpRelationsService = bpRelationsService;
+		this.jsonRetrieverService = jsonServiceFactory.createRetriever();
+	}
+
+	public JsonCreateInvoiceCandidatesResponse createInvoiceCandidates(@NonNull final JsonCreateInvoiceCandidatesRequest request)
+	{
+		final ImmutableMap<InvoiceCandidateLookupKey, JsonCreateInvoiceCandidatesRequestItem> lookupKey2Item = Maps.uniqueIndex(request.getItems(), this::createInvoiceCandidateLookupKey);
+
+		// candidates
+		final ImmutableList<ExternallyReferencedCandidate> //
+				exitingCandidates = externallyReferencedCandidateRepository.getAllBy(lookupKey2Item.keySet());
+		failIfNotEmpty(exitingCandidates);
+
+		final ImmutableList.Builder<NewManualInvoiceCandidate> candidatesToSave = ImmutableList.builder();
+		for (final Entry<InvoiceCandidateLookupKey, JsonCreateInvoiceCandidatesRequestItem> keyWithItem : lookupKey2Item.entrySet())
+		{
+			final JsonCreateInvoiceCandidatesRequestItem item = keyWithItem.getValue();
+			final NewManualInvoiceCandidate candidate = createCandidate(item);
+
+			candidatesToSave.add(candidate);
+		}
+
+		final JsonCreateInvoiceCandidatesResponseBuilder result = JsonCreateInvoiceCandidatesResponse.builder();
+		final ImmutableSet.Builder<InvoiceCandidateId> invoiceCandidateIds = ImmutableSet.builder();
+		for (final NewManualInvoiceCandidate candidateToSave : candidatesToSave.build())
+		{
+			final JsonExternalId headerId = JsonExternalIds.ofOrNull(candidateToSave.getExternalHeaderId());
+			final JsonExternalId lineId = JsonExternalIds.ofOrNull(candidateToSave.getExternalLineId());
+
+			final InvoiceCandidateId candidateId = externallyReferencedCandidateRepository.save(manualCandidateService.createInvoiceCandidate(candidateToSave));
+			invoiceCandidateIds.add(candidateId);
+			final JsonInvoiceCandidatesResponseItem responseItem = JsonInvoiceCandidatesResponseItem.builder()
+					.externalHeaderId(headerId)
+					.externalLineId(lineId)
+					.metasfreshId(MetasfreshId.of(candidateId))
+					.build();
+			result.responseItem(responseItem);
+		}
+
+		//ensure NetAmtToInvoice and other related fields are correctly populated.
+		invoiceCandBL.updateInvalid()
+				.setContext(Env.getCtx(), ITrx.TRXNAME_ThreadInherited)
+				.setTaggedWithAnyTag()
+				.setOnlyInvoiceCandidateIds(InvoiceCandidateIdsSelection.ofIdsSet(invoiceCandidateIds.build()))
+				.update();
```

Comment:
I'm not entirely sure this is wanted, but I was not able to create and enqueue for invoicing an IC from API without this. (Enqueueing failed because the IC would be changed during it. See de.metas.invoicecandidate.api.impl.InvoiceCandidateEnqueuer#enqueueSelectionInTrx )

Reply:
I see the risk of running into race-condidions while the same code is executed via async-workpackage.
I think if make a breakpoint that stops the whole program, it should turn out that the ICs are already flagged for recompute (=> `c_invoice_candidate_recompute`), at this point, right?
So you could wait for the workpackage-processor to do the work, like `de.metas.cucumber.stepdefs.StepDefUtil#tryAndWaitForItem(long, long, de.metas.cucumber.stepdefs.ItemProvider<T>)` does it.

Reply:
yup, looks like I just wasn't waiting enough in my cucumber test. removed